skimage/__init__.py

```diff
@@ -68,13 +68,38 @@
 
 """
 
-import sys
-
-
 __version__ = '0.19.0.dev0'
 
+__all__ = [
+    'color',
+    'data',
+    'draw',
+    'exposure',
+    'feature',
+    'filters',
+    'future',
+    'graph',
+    'io',
+    'measure',
+    'metrics',
+    'morphology',
+    'registration',
+    'restoration',
+    'segmentation',
+    'transform',
+    'util',
+    'viewer'
+]
+
 from ._shared.version_requirements import ensure_python_version
-ensure_python_version((3, 5))
+ensure_python_version((3, 7))
+
+from ._shared import lazy
+__getattr__, __lazy_dir__, _ = lazy.install(__name__, __all__)
```

Comment:
In this case "__all__" happens to be only a list of submodules, so this works, but I think it makes things confusing. If someone unfamiliar with the code looked at this it may appear that either `lazy.install` is able to figure out what each item in `__all__` means, or something else.

I think what would be a more intuitive way to write this code is if the first `__all__` variable was renamed to `__submodules__` or `submodules` to make it clear that it is simply a list of submodules, and not necessarily any imported attributes. Then change the `_` return to either be `__all__` or rename it to `__lazy_all__`, and so something similar like you are doing with `__dir__`.

(Or maybe, you only return `lazy.install` only returns the `__getattr__` and a `__lazy_all__` variable and it is the user's resposibility to setup `__dir__` and `__all__`, because you are kind of doing that already?)

Reply:
Personally, I don't like the name `install`, that evokes some sort of installation process to me (like apt install or running a windows installer). Obviously we can't use `lazy.import`, but are we fixed on using the `install` terminology? It feels like `lazy_import`, `lazyimport`, or something else might be better.

Reply:
Maybe `lazy.expose`?

Reply:
I think anything that doesn't re-use the module name is fine; so `lazy.expose` would work (although `expose` has a bit of a weird connotation in general).  @jarrodmillman Naming thoughts?

Reply:
I think the `__dir__` return is convenient, but it is not usable in all cases.  The naming change suggestions are sensible, I will update those.

Reply:
How about `lazy.attach`?  (I also thought about, but didn't like as much: `lazy.setup`, `lazy.manifest`, `lazy.teether`, `lazy.connect`).

Reply:
I've renamed it to `lazy.attach` for now.

Reply:
Maybe we need make PEP for Python 3.11 to add a `__call__` to modules so we can just do `lazy()` :smile:. I would like to find a better name than `attach`, but I don't have better ideas at the moment. In this spirit of brainstorming I'll just throw out ideas, some are probably bad: `init`, `new`, `module`, `define`, `delayed_import`, `inject`, `load`. I'm actually kind of partial to `lazy.load(...)`, 4 characters, reasonably clear, alteration, no duplication.

Reply:
Except that lazy.load is now used for lazy loading of external models, so then we'd need an alternative name for that function.

I used 'require' originally after Javascript and emacs.

Reply:
I like `require`.

Reply:
Currently, there is a clear distinction between the functionality of the two functions: `attach` configures a bunch of things, whereas `load` does an actual import.  If we can find other words to convey that, we can revisit.

Reply:
I like the `__call__` on the module idea, btw :D

Reply:
The [difference between](https://stackoverflow.com/questions/19198166/whats-the-difference-between-a-module-and-a-library-in-python) `library`, `package` and `module` is subtle, but I think it is useful when talking about this kind of namespace management.  The goal of the `__init__.py` module in a `library` is to create a `package` with a namespace that includes `subpackages` and attributes from `subpackages`. Technically `module` is a Python object, though it is often used to refer to a file containing the code for this object.  To help organize modules and provide a naming hierarchy, Python has a concept of packages.  As a rough analogy, modules are like files while packages are like directories. The `__init__.py` module in the package determines the namespace for the package. The term subpackage is used similarly to the term subdirectory for file systems.  I'm thinking that a module is defined in a single file, a package is held in a directory with an `__init__.py` file to load more than one module.  That package could then be a subpackage of another package, etc.  In the current code the term `module` is used to refer to many different kinds of objects -- though they are technically all python objects, the distinction between code from a single file and code from a directory is useful here.

module <-> python file
package <-> python directory (with __init__.py file)
subpackage <-> python directory within another python directory (like subdirectory)

So, for the name of the `attach` function, I like `lazy.package_structure`.  That is, this function constructs the package structure for this package. We give it a name for the package, a list of subpackages and attributes to include in the namespace and it lazily adds those entries to the namespace. 

The name is definitely longer than `attach`, and it is not a verb, so maybe there is a better name. But by not being a verb, it makes clear that this is where we set up the structure for the namespace of this package.

Reply:
I agree that the distinction is important, but I am not sure that `lazy.package_structure` captures that correctly.  `lazy.attach` is, indeed, used to set up members of a module, not a package, and can be used by many modules and submodules *within* a package.  `package_structure` would refer to the overarching package (thus, multiple modules / submodules).